test/Test/BaseDAO.hs

```diff
@@ -0,0 +1,344 @@
+      , odContract = baseDAOContract p
+      }
+  dao <- originate originateData
+  pure ((owner1, operator1), (owner2, operator2), dao, admin)
+
+test_BaseDAO :: TestTree
+test_BaseDAO = testGroup "tests to check BaseDAO contract functionality"
+  [ testGroup "Operator:"
+      [ testCase "allows zero transfer from non-existent operator" $
+          nettestTestExpectation zeroTransferScenario
+      , testCase "aborts if there is a failure (due to low balance)" $
+          nettestTestExpectation lowBalanceScenario
+      , testCase "aborts if there is a failure (due to non existent source account)" $
+          nettestTestExpectation noSourceAccountScenario
+      , testCase "aborts if there is a failure (due to bad operator)" $
+          nettestTestExpectation badOperatorScenario
+      , testCase "accepts an empty list of transfers" $
+          nettestTestExpectation emptyTransferListScenario
+      , testCase "validates token id" $
+          nettestTestExpectation validateTokenScenario
+      , testCase "cannot transfer foreign money" $
+          nettestTestExpectation noForeignMoneyScenario
+      ]
+  , testGroup "Owner:"
+      [ testCase "aborts if there is a failure (due to low balance)" $
+          nettestTestExpectation lowBalanceOwnerScenario
+      , testCase "validates token id" $
+          nettestTestExpectation validateTokenOwnerScenario
+      , testCase "cannot transfer foreign money" $
+          nettestTestExpectation noForeignMoneyOwnerScenario
+      , testCase "balanceOf request validates token id" $
+          nettestTestExpectation balanceOfOwnerScenario
+      ]
+  , testGroup "Admin:"
+    [ testCase "transfer tokens from any address to any address" $
+        nettestTestExpectation adminTransferScenario
+    ]
+  ]
+
+zeroTransferScenario :: (Monad m) => NettestImpl m -> m ()
+zeroTransferScenario = uncapsNettest $ do
+  nonexistent :: Address <- newAddress "nonexistent"
+  ((owner1, _), _, dao, _) <- originateBaseDAO SampleFA2.defaultPermissionDescriptor
+  let
+    transferData = TransferData
+      { tdFrom = AddressResolved nonexistent
+      , tdTo = addressResolved dao
+      , tdAmount = toMutez 0
+      , tdEntrypoint = DefEpName
+      , tdParameter = FA2.Transfer [
+          ( #from_ .! nonexistent
+          , #txs .! [ (#to_ .! owner1, (#token_id .! 0, #amount .! 0)) ]
+          ) ]
+      }
+  transfer transferData
+
+lowBalanceScenario :: (Monad m) => NettestImpl m -> m ()
+lowBalanceScenario = uncapsNettest $ do
+  ((owner1, op1), (owner2, _), dao, _) <- originateBaseDAO SampleFA2.defaultPermissionDescriptor
+  let
+    transferData = TransferData
+      { tdFrom = AddressResolved op1
+      , tdTo = addressResolved dao
+      , tdAmount = toMutez 0
+      , tdEntrypoint = DefEpName
+      , tdParameter = FA2.Transfer [
+          ( #from_ .! owner1
+          , #txs .! [ (#to_ .! owner2, (#token_id .! 0, #amount .! 200)) ]
+          ) ]
+      }
+  transfer transferData `expectFailure`
```

Comment:
Oh, I wanted to write some comments yesterday, but forgot, and now there are more tests to fix, sorry :pensive: 

1. There are special cases of `transfer` adapted to smart contract calls where amount is 0: `call` and `callFrom`. I think it's better to use these functions, their use is smaller (no need to construct a large `TransferData` record).
2. You are calling the default entrypoint and construct a value of the whole parameter type. It should work but has two drawbacks. First, it forces us to update all these tests when parameter type changes. Currently it has only FA2 methods, but later it will have more and probably `FA2.Parameter` will be wrapped into another constructor. So we'll have to add something like `CallFA2 $` here. If you specify the entrypoint name explicitly, the parameter will always be just `TransferParams`. Second, it makes logs a bit less clear because in this case the parameter is a large value from which it's not clear which entrypoint is called. So it's better to specify specific entrypoint name in tests.

Reply:
That makes sense, thanks for explanation, Ivan.

I've updated to use `callFrom` instead, you can check it.